### DISCUSSION TITLE
User data IQ width and compression method mismatch

### DISCUSSION THREAD

**Discussion by wei-lou** [srsran/srsRAN_Project#379] [General Help] [answered]
I have configured the IQ width as 9 and compression method bfp. However from captured pcap, it shows IQ width as 16 bits and no compression.

  compr_method_ul: bfp                                            
  compr_bitwidth_ul: 9                                            
  compr_method_dl: bfp                                            
  compr_bitwidth_dl: 9                                          
  compr_method_prach: bfp                                 
  compr_bitwidth_prach: 9                                   
  enable_ul_static_compr_hdr: true                               
  enable_dl_static_compr_hdr: true                                
  iq_scaling: 5.5 

    C-Plane Section Type 1, Downlink, Frame: 85, Subframe: 7, Slot: 0, StartSymbol: 0
        1... .... = Data Direction: Downlink (1)
        .001 .... = Payload Version: 1
        .... 0000 = Filter Index: standard channel filter (0)
        Frame ID: 85
        0111 .... = Subframe ID: 7
        .... 0000 00.. .... = Slot ID: 0
        ..00 0000 = Start Symbol ID: 0
        [RefA: 85-7-0]
        Number of Sections: 1
        Section Type: Most DL/UL radio channels (1)
        0000 .... = User Data IQ width: I and Q are each 16 bits wide (0) (16 bits)
        .... 0000 = User Data Compression Method: No compression (0)

[gnb_ru_fuji_tdd_n48_40mhz_2x2_23101BM.txt](https://github.com/srsran/srsRAN_Project/files/13730362/gnb_ru_fuji_tdd_n48_40mhz_2x2_23101BM.txt)

[IQ_width_Compress.zip](https://github.com/srsran/srsRAN_Project/files/13730429/IQ_width_Compress.zip)

What haveI missed?

**Comment by andrepuschmann** [comment]
Are you sure this PCAP is from a run with this config? Can you please share the console messages when you start the gNB. It prints the OFH config.

**Reply by wei-lou** [reply]
Hi @andrepuschmann. yes, that's the match pcap with the configure file. Here are the console screen display. I also recapture the pcap and gnb.log. The pcap indicated no compression and 16bit IQ width.

user@srsran:~$ sudo gnb -c srsRAN_Project/configs/gnb_ru_fuji_tdd_n48_40mhz_4x4_23101BM.yml
```

The PRACH detector will not meet the performance requirements with the configuration {Format 0, ZCZ 0, SCS 1.25kHz, Rx ports 1}.

--== srsRAN gNB (commit 55c984b55) ==--

Connecting to AMF on 192.168.70.210:38412
Initializing Open Fronthaul Interface sector=0: ul_comp=[BFP,9], dl_comp=[BFP,9], prach_comp=[BFP,9] prach_cp_enabled=true, downlink_broadcast=false.
Warning: Configured PRACH occasion collides with PUCCH RBs ([0..1) intersects [0..3)). Some interference between PUCCH and PRACH is expected.
Warning: Configured PRACH occasion collides with PUCCH RBs ([0..1) intersects [0..3)). Some interference between PUCCH and PRACH is expected.
Cell pci=48, bw=40 MHz, dl_arfcn=643464 (n48), dl_freq=3651.96 MHz, dl_ssb_arfcn=642432, ul_freq=3651.96 MHz

==== gNodeB started ===
Type <t> to view trace
q
Stopping ..

``` 

[gnb_run04.log](https://github.com/srsran/srsRAN_Project/files/13743726/gnb_run04.log)

[srsran_BM_n48_40MHz_DU_run04.pcap.gz](https://github.com/srsran/srsRAN_Project/files/13743727/srsran_BM_n48_40MHz_DU_run04.pcap.gz)

[gnb_ru_fuji_tdd_n48_40mhz_4x4_23101BM.txt](https://github.com/srsran/srsRAN_Project/files/13743740/gnb_ru_fuji_tdd_n48_40mhz_4x4_23101BM.txt)

**Comment by mmailand** [comment]
As far as I understand that behavior is correct, the udCompHdr value is not used in the C-Plane in the downlink direction and therefore it is set to all zeros. See the O-RAN.WG4.CUS.0-R003-v13.00 specification section 7.5.2.10 for more details:

_This means the udCompHdr in a C-Plane message only has meaning for UL U-Plane data (dataDirection = 0b) and should be set to 0x00 when dataDirection = 1b (DL data), and the receiving O- RU shall ignore the received udCompHdr value_ 

The second point is that static data format and compression is used and therefore no udCompHdr is needed, section 7.5.2.10:

_When static data format and compression is used (see clause 8.1.2) then udCompHdr is not needed in any C-Plane message and should be set to 0x00, and the receiving O-RU shall ignore the received udCompHdr value._ 

There also won't be any udCompHdr values in the U-Plane because of the static format and compression, section 8.3.3.13:

_This field shall be absent from U-Plane messages when the static IQ format and compression method is configured via the M-Plane_

**Reply by wei-lou** [reply]
@mmailand thanks for the reply.

I wasn't fully understanding your explanation as well as the WG4.CUS.0-R003-v13.0 section 7.5.2.10. I assumed that section was for the header compression rather than U-Plane data compression (Section 8.1.2).  While executing the gnb program, the console display indicated DL, UL and PRACH compression were BFP and IQ data bit 9. Those compression and width are for the U-Plane data compression, right? But the srs pcap indicates no compression and iq width bit 16.

I compared the srs pcap with other cudu pcap, there is a difference.

PCAP from srs cudu
```
Time           Source                Destination           Protocol Length Start Symbol ID IMSI       RefA       Payload Size c_eAxC_ID  Info                                                            Exponent   Data Direction
0.000000       IntelCor_ed:d8:68     CigShang_42:82:df     O-RAN-FH-C 64     0                          191-5-0    20           0:0:0:0    C-Plane, Type: 1 (Most channels), Id: 0 (PRB: 0-105)                       Downlink

Frame 1: 64 bytes on wire (512 bits), 64 bytes captured (512 bits)
Ethernet II, Src: IntelCor_ed:d8:68 (3c:fd:fe:ed:d8:68), Dst: CigShang_42:82:df (60:57:47:42:82:df)
802.1Q Virtual LAN, PRI: 0, DEI: 0, ID: 202
evolved Common Public Radio Interface
O-RAN Fronthaul CUS-C, Type: 1 (Most channels), Id: 0 (PRB: 0-105)
    ecpriRtcid (DU_Port_ID: 0, BandSector_ID: 0, CC_ID: 0, RU_Port_ID: 0)
    ecpriSeqid, SeqId: 12, SubSeqId: 0, E: 1
    C-Plane Section Type 1, Downlink, Frame: 191, Subframe: 5, Slot: 0, StartSymbol: 0
        1... .... = Data Direction: Downlink (1)
        .001 .... = Payload Version: 1
        .... 0000 = Filter Index: standard channel filter (0)
        Frame ID: 191
        0101 .... = Subframe ID: 5
        .... 0000 00.. .... = Slot ID: 0
        ..00 0000 = Start Symbol ID: 0
        [RefA: 191-5-0]
        Number of Sections: 1
        Section Type: Most DL/UL radio channels (1)
        0000 .... = User Data IQ width: I and Q are each 16 bits wide (0) (16 bits)
        .... 0000 = User Data Compression Method: No compression (0)
        Reserved: 0
    Section, Id: 0 (PRB: 0-105), Symbols: 14, BeamId: 0)
``` 

PCAP from other's cudu
```
Time           Source                Destination           Protocol Length Start Symbol ID IMSI       RefA       Payload Size c_eAxC_ID  Info                                                            Exponent   Data Direction
0.000202       NVIDIA_6c:70:70       CigShang_42:82:d9     O-RAN-FH-C 64     3                          230-5-0    20           0:0:0:0    C-Plane, Type: 1 (Most channels), Id: 2 (PRB: 151-163)                     Downlink

Frame 9: 64 bytes on wire (512 bits), 64 bytes captured (512 bits)
Ethernet II, Src: NVIDIA_6c:70:70 (48:b0:2d:6c:70:70), Dst: CigShang_42:82:d9 (60:57:47:42:82:d9)
802.1Q Virtual LAN, PRI: 7, DEI: 0, ID: 3
evolved Common Public Radio Interface
O-RAN Fronthaul CUS-C, Type: 1 (Most channels), Id: 2 (PRB: 151-163)
    ecpriRtcid (DU_Port_ID: 0, BandSector_ID: 0, CC_ID: 0, RU_Port_ID: 0)
    ecpriSeqid, SeqId: 33, SubSeqId: 0, E: 1
    C-Plane Section Type 1, Downlink, Frame: 230, Subframe: 5, Slot: 0, StartSymbol: 3
        1... .... = Data Direction: Downlink (1)
        .001 .... = Payload Version: 1
        .... 0000 = Filter Index: standard channel filter (0)
        Frame ID: 230
        0101 .... = Subframe ID: 5
        .... 0000 00.. .... = Slot ID: 0
        ..00 0011 = Start Symbol ID: 3
        [RefA: 230-5-0]
        Number of Sections: 1
        Section Type: Most DL/UL radio channels (1)
        1001 .... = User Data IQ width: Bit width of I and Q (9) (9 bits)
        .... 0001 = User Data Compression Method: Block floating point compression (1)
        Reserved: 0
    Section, Id: 2 (PRB: 151-163), Symbols: 1, BeamId: 0)
``` 

Is there any other configuration lines I need to add to the configuration file to ensure the right compression method and IQ for U-Plane data?

**Reply by mmailand** [reply]
If you want to have the udCompHdr in the U-Plane you can do that via the config option:
`enable_ul_static_compr_hdr: false`  
`enable_dl_static_compr_hdr: false`
and than adjust the different compr_bitwidth and compr_method settings as you mentioned in the your first post, e.g. for DL:
`compr_bitwidth_dl: 9`
`compr_method_dl: bfp`

Because there is no flag in the specification to indicate the presents of the udCompHdr we need to configure that in the wireshark preferences. Tick those two boxes and save the preferences. 
![wireshark](https://github.com/srsran/srsRAN_Project/assets/1290351/1029e250-cf2f-4209-ba1d-25fd3769d6d7)

After that you can see the expected udCompHdr in the U-Plane.
![udCompHdr](https://github.com/srsran/srsRAN_Project/assets/1290351/051b6acc-1644-420e-8ad7-29984f48bec1)

**Reply by wei-lou** [reply]
great! let me try that. Thank you so much for your help

**Reply by andrepuschmann** [reply]
Thanks @mmailand for helping out. As @mmailand pointed out the trouble with the OFH spec is that some config options are not type-length-value (TLV) in the messages but are always present, regardless of whether they are used or not. It is a bit counter intuitive. So when viewing the messages with Wireshark you need to make sure the dissector is configured accordingly.

**Reply by pgawlowicz** [reply, marked as answer]
@wei-lou Did the hint from @mmailand solve your issue? Could you provide your feedback?

**Reply by wei-lou** [reply]
Yes, the explanation makes sense. There was no issue here. Thanks,

**Comment by andrepuschmann** [comment]
Thanks for the feedback. Closing.
